runtime/src/main/cpp/MemorySharedRefs.cpp

```diff
@@ -102,12 +110,16 @@ void BackRefFromAssociatedObject::addRef() {
 }
 
 bool BackRefFromAssociatedObject::tryAddRef() {
+  return konan::tryOrTerminate([this]() { return tryAddRefOrThrow(); });
+}
+
+bool BackRefFromAssociatedObject::tryAddRefOrThrow() {
   // Suboptimal but simple:
-  this->ensureRefAccessible();
+  this->ensureRefAccessibleOrThrow();
   ObjHeader* obj = this->obj_;
 
   if (!TryAddHeapRef(obj)) return false;
-  this->addRef();
+  this->addRefOrThrow();
```

Comment:
Ypu may need RAII here to balance TryAddHeapRef. However, looks like it can't throw actuall after `ensureRefAccessibleOrThrow` above. Anyway, consider to clarify the point with comment or RAII or (probably better) try to avoid duplicated check in ensureRefAccessibleOrThrow an then in addRefOrThrow

Reply:
Good idea. I replaced this with `RuntimeAssert` + (crashing) `addRef` for now, but it's still suboptimal. Better yet is to use a variant of `addRef` that doesn't have any checks at all. I'm still figuring out a nicer way of doing that, considering your comment above